### PR TITLE
Small fixes and unit tests [bugfix/nd-segment-dev]

### PR DESCRIPTION
**Pull request by mlstowell**
After merging master into an obscure branch a couple of unit tests began failing. The relevant unit test and the necessary small fixes are provided here.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2163](https://github.com/mfem/mfem/pull/2163) | @mlstowell | @v-dobrev | @psocratis + @dylan-copeland | 04/14/21 | 04/27/21 | 05/04/21 | |

### CODE REVIEW

File: fem/fe.cpp [hunk 1]
```diff
@@ -12604,7 +12604,7 @@ void ND_TriangleElement::CalcCurlShape(const IntegrationPoint &ip,
 const double ND_SegmentElement::tk[1] = { 1. };
 
 ND_SegmentElement::ND_SegmentElement(const int p, const int ob_type)
-   : VectorFiniteElement(1, Geometry::SEGMENT, p, p - 1,
+   : VectorFiniteElement(1, Geometry::SEGMENT, p, p,
```
Comment (v-dobrev) [inline fem/fe.cpp:12607]: I don't think this is correct -- Nedelec `p` elements on edges have order `p-1`. For example, in the lowest order case, `p=1`, the tangential components are constants on the edges. This is also seen in the line:
```
      obasis1d(poly1d.GetBasis(p - 1, VerifyOpen(ob_type)))
```
below.

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/fe.cpp (hunk 1 above)
Comment (mlstowell) [reply fem/fe.cpp:12607]: I should have done a better job of documenting this.

The `ND_FECollection` of order `p` allocates the various `ND_*Element` classes by passing `p` to them as the order. It also returns `p` when `ND_FECollectxion::GetOrder()` is called. The `FiniteElementSpace::GetFE(int i)` method recently introduced an `MFEM_ASSERT` that checks that the `FE_Collection::GetOrder()` and `FiniteElement::GetOrder()` methods return the same integer.

Possible fixes were:
1) Disable this new check either completely or in this special case when the FiniteElement is an ND_SegmentElement.
2) Allocate an order `p+1` ND_SegmentElement in the order `p` ND_FECollection.
3) Increase the finite element order of `ND_SegmentElement` from `p-1` to `p`.
4) Perhaps something else...

Option 3 seemed to be the most consistent with the other `ND_*Element` classes which use the order perpendicular to the edge rather than along the edge.

Comment (psocratis) [reply fem/fe.cpp:12607]: Thanks @mlstowell , this makes sense. 

Since this the only special case (right?) I would vote for actually modifying correctly the `MFEM_ASSERT` for the `ND_SegmentElement` (not removing it).

File: fem/lininteg.cpp [hunk 2]
```diff
@@ -48,7 +48,7 @@ void DomainLFIntegrator::AssembleRHSElementVect(const FiniteElement &el,
       Tr.SetIntPoint (&ip);
       double val = Tr.Weight() * Q.Eval(Tr, ip);
 
-      el.CalcShape(ip, shape);
+      el.CalcPhysShape(Tr, shape);
```
Comment (dylan-copeland) [inline fem/lininteg.cpp:51]: Is it correct that this change just divides by the weight for the `INTEGRAL` map type? Why is this change needed, and why did it not matter until now? Will this change results for systems that use the `INTEGRAL` type?

Comment (mlstowell) [reply fem/lininteg.cpp:51]: Yes, this just devides by Det[J] for the `INTEGRAL` map type. Previously this integrator did not support map type `INTEGRAL` and using this combination would quietly produce the wrong answer. This is another small fix for the issue that Dan raised some time ago in #1037. It's my understanding that people generally don't use the `INTEGRAL` map type so I doubt anyone will notice. However, those who do choose to use it will find that this integrator works as expected and they won't need to track down the source of oddly scaled results.

File: fem/fe.cpp (hunk 1 above)
Comment (mlstowell) [reply fem/fe.cpp:12607]: I was thinking that the `GetOrder` methods were returning the order of convergence not the highest polynomial degree. A piecewise constant basis does converge at first order, doesn't it?

Comment (tzanio) [reply fem/fe.cpp:12607]: `GetOrder` should be the polynomial degree of the space, ping: @jakubcerveny

Comment (jakubcerveny) [reply fem/fe.cpp:12607]: The assert was added in `var-order-space-dev` to make sure the element order in the space is the same thing as the FE order. I have always been confused about the order of the lower-order entities in ND, but perhaps we can modify the check to only apply to the actual elements (`FE->GetDim() == mesh->Dimension()`), not edges or faces (`FE->GetDim() > 0`)?

Comment (mlstowell) [reply fem/fe.cpp:12607]: We may end up with a related problem in `FiniteElementSpace::GetBE(int i)` for RT spaces. This method requests a `FiniteElement` from the `FE_Collection` with the same `order` as the neighboring interior element. However, for RT spaces the boundary elements are L2 elements which, by this definition, have lower order than the interior element.

Comment (mlstowell) [reply fem/fe.cpp:12607]: Perhaps there is a need for two separate integers in these classes. One to indicate that `FiniteElement` classes which share the same value of this parameter all have compatible basis functions. And a second integer to indicate the polynomial degree so that adequate integration rules can be chosen. These are two separate concerns after all.

Comment (mlstowell) [reply fem/fe.cpp:12607]: I propose we change `FiniteElementCollection::GetOrder()` to `FiniteElementCollection::GetConvergenceOrder()` (feel free to propose other names) and we add `FiniteElement::GetConvergenceOrder()` (same name as in `FiniteElementCollection` whatever we decide for that).  Then @jakubcerveny's variable order modifications can verify that the integers returned from these new methods agree.

With this change we can continue to use `FiniteElement::GetOrder()` in the manner that @v-dobrev and @tzanio intended i.e. as a guide to selecting integration rule orders.

Ping: @dylan-copeland and @psocratis .

File: fem/lininteg.cpp (hunk 2 above)
Comment (v-dobrev) [reply fem/lininteg.cpp:51]: This can probably be optimized to avoid computing `Tr.Weight()` for `INTEGRAL` map-type -- we're now multiplying by that and then dividing by it. 😁

File: fem/fe.cpp (hunk 1 above)
Comment (psocratis) [reply fem/fe.cpp:12607]: I prefer leaving the `GetOrder()` as is (corresponding to the polynomial degree), and if we were to introduce a new one, I would prefer a name that refers to the de Rham Complex (exact sequence) order. For example on the discrete level in 3D the sequence is `W^p (grad)->Q^p (curl)->V^p (div)->Y^p` , where the symbol `p` is not neccessarily the polynomial order (i.e., for Cartesian product elements this is a multi-index)

Comment (v-dobrev) [reply fem/fe.cpp:12607]: @mlstowell or @jakubcerveny, where is this `MFEM_ASSERT`?

Comment (mlstowell) [reply fem/fe.cpp:12607]: @v-dobrev , it is in `FiniteElementSpace::GetFE(int i)`

Comment (tzanio) [reply fem/fe.cpp:12607]: I have not followed everything here, but I just wanted to add that there was a (long) discussion about `GetOrder` and `GetFaceOrder` in https://github.com/mfem/mfem/pull/1817, see e.g. here https://github.com/mfem/mfem/pull/1817/commits/2a0b624fd1cdad101a4e3bcd721d93a7e9ddcab3.

Comment (dylan-copeland) [reply fem/fe.cpp:12607]: If I understand correctly, in master, `GetOrder` is the polynomial degree, and for constant-order spaces (same order on all elements, meaning no p-refinements), `FiniteElementSpace::GetFE` asserts that `GetOrder` is the same for the `FiniteElement` and the `FiniteElementCollection`. It seems simplest not to change these functions, if possible. What is needed to fix the issue Mark found is either logic to change or bypass the assertion in this special case, or to add a new type of order to be checked in the assertion. There have been suggestions for the new type of order to be order of convergence or the de Rham complex order, and those both sound fine to me. We could have some new order function in addition to `GetOrder`, without changing `GetOrder`. Initially, the new order function would only be used in this one assertion, but it could be useful in the future in other ways.

As an alternative, if we just want to fix this assertion that all elements have the same order as the collection in some sense, one possibility, without introducing any new functions or concepts, would be to check the number of DOFs. That is, replace `FE->GetOrder() == fec->GetOrder()` with
`FE->GetDof() == fec->GetNumDof(mesh->GetElementGeometry(i), fec->GetOrder())`
In the `ND_SegmentElement` case, this would be checking `p == p` although the results of `GetOrder` are different.

Comment (v-dobrev) [reply fem/fe.cpp:12607]: Dylan, there's one issue with comparing the number of dofs: `fec->GetNumDof` and the method it uses, `DofForGeometry`, return the number of **interior** dofs. On the other hand, `FE->GetDof` returns all dofs including the ones on the boundary of the element.

One simple solution for the issue could be to replace https://github.com/mfem/mfem/blob/913361dd0db723d07e9c7492102f8b887acabc48/fem/fe_coll.cpp#L2520-L2523

with
```c++
 ND_FECollection::ND_FECollection(const int p, const int dim, 
                                  const int cb_type, const int ob_type) 
    : FiniteElementCollection(dim > 1 ? p : p-1) 
 { 
```
which sets the **correct order** based on `dim`. This will prevent us from using the collection `ND_FECollection(p,dim)`, `dim=2,3` with 1D meshes which has been valid so far (before we introduced variable orders). In other words, we'll be breaking backward compatibility but, in this case, that is probably acceptable.

I think it is important to remember (and we should document this in the base class `FiniteElementCollection`) that the number returned by `GetOrder` and the `p` parameter in the variable-order methods `Clone`, `GetFE`, `GetNumDof`, and `GetDofOrdering` are generally NOT the same as the `p` parameter used the constructors of derived classes. Also, the `p` parameter in the variable-order methods `GetFE`, `GetNumDof`, and `GetDofOrdering` is generally NOT the polynomial order for the element corresponding to the `geom` parameter.

File: fem/lininteg.cpp (hunk 2 above)
Comment (mlstowell) [reply fem/lininteg.cpp:51]: @v-dobrev, you are correct of course. I prefer using `CalcPhysShape` because it hides details of the basis functions from the integrator which I find makes the integrator cleaner and easier to follow. For me this is worth the loss of this particular optimization. However, I'll be happy to change this if you think it is important to do so.

Comment (v-dobrev) [reply fem/lininteg.cpp:51]: I don't think it is critical to do the optimization, so you I'm okay to leave it as it is.

File: fem/fe.cpp (hunk 1 above)
Comment (v-dobrev) [reply fem/fe.cpp]: I pushed the above fix -- it should fix the unit test failure.

I also added some comments, as I was suggesting above -- @jakubcerveny you may want to take a look at these too.

@mlstowell, @dylan-copeland, @psocratis, or anyone else, let me know if this solution does not look right to you.

File: tests/unit/fem/test_domain_int.cpp [hunk 3]
```diff
@@ -0,0 +1,671 @@
...
+            delete fec;
+         }
+      }
+
+      delete mesh;
+   }
+}
+
+#ifdef MFEM_USE_MPI
+#
```
Comment (dylan-copeland) [inline tests/unit/fem/test_domain_int.cpp:246]: Typo?

File: tests/unit/fem/test_domain_int.cpp [hunk 4]
```diff
@@ -0,0 +1,671 @@
...
+   int my_rank;
+   MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
+
+   int order = 3;
+
+   for (int mt = (int)MeshType::SEGMENT;
+        mt <= (int)MeshType::MIXED3D8; mt++)
+   {
+      Mesh *mesh = GetMesh((MeshType)mt);
+      int  dim = mesh->Dimension();
```
Comment (dylan-copeland) [inline tests/unit/fem/test_domain_int.cpp:266]: Two spaces.

**Review by dylan-copeland — APPROVED** [review]

**Review by jakubcerveny — APPROVED** [review]

**Review by psocratis — APPROVED** [review]

**Comment by v-dobrev** [comment]
@mlstowell, just to make sure before I merge this in `next`: are you okay with the changes I pushed?

**Comment by mlstowell** [comment]
> @mlstowell, just to make sure before I merge this in `next`: are you okay with the changes I pushed?

Honestly, I don't see the point of `FiniteElementCollection::GetOrder` and I would prefer a different method which returned something that could be used to identify a family of related `FiniteElement`s which are compatible with each other. As it stands the user must know that an `ND_SegmentElement` of order `p` is compatible with the other `ND_FiniteElement`s of order `p+1`.

That being said, I am not interested in arguing the point. We can go ahead and merge this whenever you think it's ready.

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
LGTM
